tools/asdf/src/proto.rs

```diff
@@ -0,0 +1,251 @@
+        .into());
+    };
+
+    if git {
+        return Ok(repo.trim().to_string());
+    }
+
+    let repo = repo.split(".git").next();
+    let Some(repo) = repo else {
+        return Err(PluginError::Message(String::from("Repository not found in downloaded file"))
+        .into());
+    };
+
+    Ok(format!("{}/bin/{script}", get_raw_github_url(repo)))
+}
+
+/// Returns the real path of the script
+fn get_script(script: &str) -> FnResult<String> {
+    let id = get_plugin_id()?;
+
+    let virtual_script_path = virtual_path!(format!("/proto/temp/{id}"));
+    let script_url = get_script_url(script, true)?;
+    let repo_dir = virtual_script_path.join("asdf");
+    
+    fs::create_dir_all(virtual_script_path.parent().unwrap())?;
+    exec_command!("rm", ["-rf", repo_dir.real_path().unwrap().to_str().unwrap()]);
+    exec_command!("git", ["clone", &script_url, repo_dir.real_path().unwrap().to_str().unwrap()]);
+
+    let real_script_path = virtual_script_path.join("asdf").join("bin").real_path().unwrap().into_os_string().into_string().unwrap();
+    Ok(format!("{real_script_path}/{script}"))
+}
+
+/// Check if the requested tool version is available in the asdf versions.
+fn check_version(version: &str) -> FnResult<()> {
+    let versions = exec_command!("bash", [get_script("list-all")?]).stdout;
+    let versions = versions.split_whitespace();
+
+    if version == "latest" {
+        return Ok(());
+    }
+
+    for v in versions {
+        if v == version {
+            return Ok(());
+        }
+    }
+
+    Err(PluginError::Message(format!("Version {version} not found in asdf repository")).into())
+}
+
+/// Set the environment variables to be used by asdf scripts
+fn set_asdf_env_vars(version: &str) -> FnResult<()> {
+    host_env!("ASDF_INSTALL_TYPE", "version");
+    host_env!("ASDF_INSTALL_VERSION", version);
+
+    let asdf_bin = virtual_path!("/proto/bin/asdf");
+    host_env!("ASDF_INSTALL_PATH", asdf_bin.real_path().unwrap().to_str().unwrap());
+    asdf_bin.real_path().unwrap().to_str().unwrap();
+    host_env!("ASDF_DOWNLOAD_PATH", asdf_bin.real_path().unwrap().to_str().unwrap());
+    // asdf needs a bin directory to install the tools
+    let _ = fs::create_dir_all(asdf_bin.join("bin"));
+
+    let cores = exec_command!("nproc").stdout;
+    host_env!("ASDF_CONCURRENCY", cores);
+
+    Ok(())
+}
+
+#[plugin_fn]
+pub fn register_tool(Json(input): Json<ToolMetadataInput>) -> FnResult<Json<ToolMetadataOutput>> {    
+    Ok(Json(ToolMetadataOutput {
+        name: input.id,
+        type_of: PluginType::Language,
+        minimum_proto_version: Some(Version::new(0, 44, 7)),
+        plugin_version: Version::parse(env!("CARGO_PKG_VERSION")).ok(),
+		config_schema: Some(schematic::SchemaBuilder::generate::<AsdfPluginConfig>()),
+        ..ToolMetadataOutput::default()
+    }))
+}
+
+#[plugin_fn]
+pub fn download_prebuilt(Json(_): Json<DownloadPrebuiltInput>) -> FnResult<Json<DownloadPrebuiltOutput>> {
```

Comment:
So there's a hidden API that may work better than `download_prebuilt` + `unpack_archive`, and it's `native_install`. It's kind of an escape hatch, and is how rust works:

https://github.com/moonrepo/plugins/blob/master/tools/rust/src/proto.rs#L102

Reply:
Alternatively, I'm landing a "build from source" API that also could work here (especially the builder stuff): https://github.com/moonrepo/proto/pull/701

Here's some examples: https://github.com/moonrepo/plugins/pull/29